### PR TITLE
Feature/maroon 465 reusable ode solver

### LINKED ISSUE #465: Reusable 3D ODE solver
***Describe the issue***
Create a 3d ODE solver as reusable script/prefab

### PR DESCRIPTION
**Pull request by jonathanplatzer**
Implementation of a reusable solver for ODEs describing motion in 3D.
Closes #465

### CODE REVIEW

**Comment by jonathanplatzer** [comment]
Hello, first and foremost I would like to apologize for taking so long with this PR and the implementation in general and hope that my contribution is still of value to the project and not too byzantine in terms of code quality or the name of things.

This is now what I would consider near feature complete in terms of the reusable ode solver code, but not so much in terms of the code of the 3D Motion Simulation Experiment, which is kind of a mess, and could be simplified. I started working on simplifying it, as can be seen in the `ParameterLoader` and `MotionCalculation` classes, but I think this is out of scope for this pull request and I am not sure if the modification of the experiment here to just ***use*** the reusable solver is already to much and should be done in another pull request.

**Comment by jonathanplatzer** [comment]
I think this is it, this PR is at a point where it can be reviewed and (hopefully) merged.

First the reusable ODE solver is finished, maybe only the names need some work, but this is trivial, I don't expect any changes to the way the solver fundamentally works and is structured.

Second the 3D Motion Simulation experiment uses the ODE solver for calculations now and has been improved quite a bit, including:
- Started cleaning up code and simplifying it
- Hide the description panel as it is not currently in use (closes #485)
- Tackling #466 by utilizing a `LineRenderer` for trajectories, rendering the axes right from the start, updating the simulation when the input parameters were edited
- Reset now resets only the simulation part, not the whole UI state and everything
- Arbitraty expressions can be supplied by including them in the `exprs` json field, a working example of this is the new `Rocket Launch` preset, which includes a custom implementation of the Heaviside Step Function
https://github.com/GameLabGraz/Maroon/blob/25dd09a2f6c66188bb175f69a262185be4d61d79/unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/Resources/RocketLaunch.json#L11-L13
- Two new experiments are included, `Rocket Launch` and `Coupled Oscillators`, which were copied nearly verbatim from Peter Hadleys ODE solver, so #467 could be closed

File: unity/Assets/Maroon/reusableScripts/Physics/Motion/MotionEntity.cs [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        internal State current_state;
+
+        private Dictionary<String, Expression> _exprs = new();
+        private Dictionary<String, double> _params = new();
+
+        private List<State> _state;
+        private State _initial_state;
+
+        private double _dt;
+        private bool _isInitialized = false;
```
Comment (FlorianGlawogger) [inline unity/Assets/Maroon/reusableScripts/Physics/Motion/MotionEntity.cs]: is this variable necessary?

File: unity/Assets/Maroon/reusableScripts/Physics/Motion/State.cs
Comment (FlorianGlawogger) [inline unity/Assets/Maroon/reusableScripts/Physics/Motion/State.cs:1]: Could be confused with a StateMachine state maybe? What do you think about renaming it to "MotionState"?

File: unity/Assets/Maroon/reusableScripts/Physics/Motion/Tests.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
+﻿using UnityEngine;
+
+namespace Maroon.Physics.Motion
+{
+    public class Tests
```
Comment (FlorianGlawogger) [inline unity/Assets/Maroon/reusableScripts/Physics/Motion/Tests.cs]: This class should be named somehow differently, maybe MotionTests or something else so that it's clear it's related to Motion

File: unity/Assets/Maroon/reusableScripts/Physics/Motion/Tests.cs
Comment (FlorianGlawogger) [inline unity/Assets/Maroon/reusableScripts/Physics/Motion/Tests.cs:1]: IMO and if I understand correctly, these 3 tests could all be UnitTests (EditMode tests), then it would be ensured that the code works correctly even if something is changed in the future and testing would be also be automatically, thus we would immediately know if someone were to break something.

(instead of PrintData you'd need to assert the actual values with the expected values)

File: unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/3DMotionSimulation.pc.unity
Comment (FlorianGlawogger) [inline unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/3DMotionSimulation.pc.unity:1]: When starting, the default params, the camera works fine, when switching then to e.g. "Satellite" and then back to Default the camera does not work correctly anymore (Camera can zoom out too far in e.g.).

File: unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/Resources/RocketLaunch.json
Comment (FlorianGlawogger) [inline unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/Resources/RocketLaunch.json:1]: background could maybe be something different that fits a rocket launch?

File: unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/Scripts/CoordSystem.cs
Comment (FlorianGlawogger) [inline unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/Scripts/CoordSystem.cs:1]: private string _text; is unused I think?

Comment (FlorianGlawogger) [inline unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/Scripts/CoordSystem.cs:1]: private List<GameObject> _objects = new List<GameObject>(); is unused as well I think?

File: unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/Scripts/MotionCalculation.cs [hunk 3]
```diff
@@ -3,73 +3,23 @@
 using NCalc;
```
Comment (FlorianGlawogger) [inline unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/Scripts/MotionCalculation.cs]: NCalc is unused and could be removed

File: unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/Scripts/ParameterLoader.cs [hunk 4]
```diff
@@ -32,47 +28,42 @@ public class Parameters
         public float Vy;
         public float Vz;
 
+        public string m;
+        public string fx;
+        public string fy;
+        public string fz;
+
+        public Dictionary<string, string> exprs = new Dictionary<string, string>();
```
Comment (FlorianGlawogger) [inline unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/Scripts/ParameterLoader.cs]: To be honest, I am not a big fan of abbreviating variable names, E.g. writing` expressions` instead of `exprs` doesn't take that much longer with auto-completion of most IDEs, and code would become more readable.

File: unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/Scripts/SkyboxController.cs [hunk 5]
```diff
@@ -76,6 +76,5 @@ public static SkyboxController Instance
     /// </summary>
     public void ResetObject()
     {
-        SetExperimentRoom();
     }
```
Comment (FlorianGlawogger) [inline unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/Scripts/SkyboxController.cs]: If this empty method is unused, and you also removed IResetObject inheritance, then I think we can remove the whole method.

**Review by FlorianGlawogger — CHANGES_REQUESTED** [review]
Looks very nice, and works well, this is a great improvement compared to the state of the experiment before your PR. I just noticed a few unused variables, an empty method, and other similar minor things that could maybe be improved. I think if those get fixed the PR can be merged 👍

File: unity/Assets/Maroon/reusableScripts/Physics/Motion/MotionEntity.cs (hunk 1 above)
Comment (jonathanplatzer) [reply unity/Assets/Maroon/reusableScripts/Physics/Motion/MotionEntity.cs]: Which one do you mean? `_isInitialized` is not used, i think it is a remnant of an old idea, i'm going to remove it.

File: unity/Assets/Maroon/reusableScripts/Physics/Motion/State.cs
Comment (jonathanplatzer) [reply unity/Assets/Maroon/reusableScripts/Physics/Motion/State.cs:1]: Ah, thanks, good suggestion

File: unity/Assets/Maroon/reusableScripts/Physics/Motion/Tests.cs (hunk 2 above)
Comment (jonathanplatzer) [reply unity/Assets/Maroon/reusableScripts/Physics/Motion/Tests.cs]: These tests are still there from the prototyping phase, i am actually not too sure if they would be very useful to keep.
I just left them in here for documentation purposes, but now that there are some real uses in the motion experiment I think this file can be deleted.
What do you think?

File: unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/3DMotionSimulation.pc.unity
Comment (jonathanplatzer) [reply unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/3DMotionSimulation.pc.unity:1]: I now just don't update the camera limits at all and keep them at the defaults, because there is no real value in zooming farther out than with the default camera...

File: unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/Resources/RocketLaunch.json
Comment (jonathanplatzer) [reply unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/Resources/RocketLaunch.json:1]: I tried to find some background thats more fitting, but haven't found any skybox which shows a rocket launchpad or something along these lines, do you think the grass field from the ball experiment is good enough?

File: unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/3DMotionSimulation.pc.unity
Comment (FlorianGlawogger) [reply unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/3DMotionSimulation.pc.unity:1]: If I select the Satellite parameters though, and want to view it from the top; I can never see the whole motion path of the satellite (see image below)
grafik

so IMO zooming further out in the satellite scenario so that you can see the whole trajectory from above would be nice

File: unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/Resources/RocketLaunch.json
Comment (FlorianGlawogger) [reply unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/Resources/RocketLaunch.json:1]: Yes, the grass field sounds nice 👍 
Could you also please use a rocket model for the RocketLaunch parameters? E.g. the red one from this pack here would be nice [https://assetstore.unity.com/packages/3d/vehicles/space/cartoon-low-poly-spaceship-287099](https://assetstore.unity.com/packages/3d/vehicles/space/cartoon-low-poly-spaceship-287099) (But make sure when you import it, that you only add the files needed, i.e. not all three rockets, and only the materials and prefabs absolutely needed, otherwise 87MB for this one model would be too much probably

File: unity/Assets/Maroon/reusableScripts/Physics/Motion/Tests.cs (hunk 2 above)
Comment (FlorianGlawogger) [reply unity/Assets/Maroon/reusableScripts/Physics/Motion/Tests.cs]: I talked with Michael, and he agreed you can delete the Tests.cs file if it's not needed.

**Review by FlorianGlawogger — CHANGES_REQUESTED** [review]
Thanks for the changes, looks great; please see my responses to the comments above. I.e. I think there are now 3 things left: delete Tests.cs file, re-add and fix camera zoom, and change rocket launch background to grass field and add rocket model. 👍
and please merge with the develop branch (Github Actions tests also need to pass)

File: unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/3DMotionSimulation.pc.unity
Comment (jonathanplatzer) [reply unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/3DMotionSimulation.pc.unity:1]: This should work now

File: unity/Assets/Maroon/reusableScripts/Physics/Motion/Tests.cs (hunk 2 above)
Comment (jonathanplatzer) [reply unity/Assets/Maroon/reusableScripts/Physics/Motion/Tests.cs]: I removed it

File: unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/Resources/RocketLaunch.json
Comment (jonathanplatzer) [reply unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/Resources/RocketLaunch.json:1]: I added the red rocket, i still has 37MB, is this alright or is it still too much?
I could try to scale down the textures, I think 2048x2048 is a bit excessive for this use case, but I don't know if that is allowed regarding the license of the asset

**Comment by jonathanplatzer** [comment]
Ok, this is rather weird, before merging develop only the playmode tests were failing, but now after the fact also the WebGL build fails... Also the error log is rather unhelpful, I searched for errors that my code could have caused, but haven't found anything that seems related

edit:
Ok, seems like the WebGL build had just some hiccup, i re-run it and it passed. Also some good news, I figured out which test case failed, and of course it is caused by my code. 
Turns out searching for "error" in the log is not that helpful, but when while looking through all occurrences of "failed" i found the error. It seems like while restructuring the 3D Motion Experiment I inadvertently left out/removed some NullReferenceException handling.

File: unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/3DMotionSimulation.pc.unity
Comment (FlorianGlawogger) [inline unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/3DMotionSimulation.pc.unity:1]: If I select Coupled Oscillators, I get a NullReferenceException:
grafik

File: unity/Assets/Maroon/reusablePrefabs/CartoonSpaceship/Textures/RedShip_Normal.png
Comment (FlorianGlawogger) [inline unity/Assets/Maroon/reusablePrefabs/CartoonSpaceship/Textures/RedShip_Normal.png:1]: Since all of the textures combined are 37MB, maybe it's possible we can compress them/reduce size? I know I suggested this asset, so I'm sorry this is now again a bit of work to do; but maybe you could either scale down the textures to maybe 256x256 and see if it still looks fine? Maybe we can fit all textures within a ~2MB limit or something like that? I think since the spaceship will not be viewed from close up that should still be fine?
I talked with Michael, and he said otherwise you can also delete e.g. the normal maps, or any other textures besides the color textures depending on how scaling down the textures goes. If it's too much work required, you can also delete the textures and just assign a red material to the spaceship body, and a gray-ish metallic material to the landing gear.

**Review by FlorianGlawogger — CHANGES_REQUESTED** [review]
Hi, thanks for your changes again 👍 the zooming out part now works very well! :) 
I juts noticed there's a NullReferenceError when selecting the Coupled Oscillators parameters, and also maybe you can try to reduce the file size of the rocket model? (Modifying assets from the Unity Asset store is allowed) (please see my other comment regarding that)
Otherwise the PR looks very solid now, and I think when these 2 issues are fixed, I think we'll finally be able to merge it 😄 (sorry again about the long review process)

File: unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/3DMotionSimulation.pc.unity
Comment (jonathanplatzer) [reply unity/Assets/Maroon/scenes/experiments/3DMotionSimulation/3DMotionSimulation.pc.unity:1]: Ah, good catch, i wanted to make background and particle optional parameters in the JSON experiment files, which worked fine at first, but it seems there was a regression. I think everything should be fixed by using null-conditional when referencing the optional fields

File: unity/Assets/Maroon/reusablePrefabs/CartoonSpaceship/Textures/RedShip_Normal.png
Comment (jonathanplatzer) [reply unity/Assets/Maroon/reusablePrefabs/CartoonSpaceship/Textures/RedShip_Normal.png:1]: I have removed every texture besides the albedo textures, and resized them to a more reasonable 1024x1024. Now the whole thing is a much more reasonable 350KB instead of 37MB, and it looks exactly the same, so that's that...

**Review by FlorianGlawogger — APPROVED** [review]
Rocket model still looks good and the new file size is much more reasonable, coupled oscillators parameter preset now also works again!
tldr: Looks good now, thanks 👍
